File: tests/Suscribe.ts
import * as anchor from "@project-serum/anchor";
import { Program } from "@project-serum/anchor";
import { MonthlySubscriptionSolana } from "../target/types/monthly_subscription_solana";
import { Enterprise } from "/mnt/c/Users/Mateo/monthly_subscription_Solana./monthly_subscription_Solana./tests/Accounts";
import { PublicKey } from '@solana/web3.js';
const { SystemProgram } = anchor.web3;

describe("Subscribe", async () => {
    const provider = anchor.AnchorProvider.env();
    anchor.setProvider(provider);
    const wallet = provider.wallet as anchor.Wallet;
    const program = anchor.workspace.MonthlySubscriptionSolana as Program<MonthlySubscriptionSolana>;
  it('Creating subscription', async () => {
    const Account = await program.account.enterpriseData.fetch(Enterprise)
    const [userData, _bump] = await PublicKey.findProgramAddress(
      [
        Enterprise.toBuffer(),
        wallet.publicKey.toBuffer(),
      ],
      program.programId
    )
    const tx = await program.methods.suscribe(
      "Mateo",
      "Lafalce"
    ).accounts({
        enterpriseData: Enterprise,
        userData: userData,
        from: wallet.publicKey,
        stake: Account.authority,
        user: wallet.publicKey,
        systemProgram: SystemProgram.programId
      }).rpc();
      console.log("---------------------------------------------")
      console.log("PDA: ", userData.toBase58())
      console.log("---------------------------------------------")
      console.log("Tx: ", tx);
      console.log("---------------------------------------------")
  });
});

File: test_in_rust/functions/suscribe.rs
use anchor_client::{
    anchor_lang::system_program,
    solana_sdk::{
        pubkey::Pubkey,
        signature::{Keypair, Signature},
        signer::Signer,
    },
    Program,
};
use anyhow::Result;
use monthly_subscription_solana::state::EnterpriseData;

pub fn suscribe(
    program: &Program,
    name: String,
    lastname: String,
    enterprise_data: Pubkey,
) -> Result<()> {
    let user_data: Keypair = Keypair::new();
    let account: EnterpriseData = program.account(enterprise_data)?;
    let tx: Signature = program
        .request()
        .accounts(monthly_subscription_solana::accounts::Suscribe {
            enterprise_data,
            user_data: user_data.pubkey(),
            from: program.payer(),
            stake: account.authority,
            user: program.payer(),
            system_program: system_program::ID,
        })
        .signer(&user_data)
        .args(monthly_subscription_solana::instruction::Suscribe { name, lastname })
        .send()?;
    println!("Tx: {}", tx);
    Ok(())
}
